### PR TITLE
[vcpkg] Add vcpkg_check_features

### PR DESCRIPTION
**Pull request by myd7349**
When working on https://github.com/microsoft/vcpkg/pull/4979, I realized this pattern has been repeated too much:
```cmake
if(<feature> IN_LIST FEATURES)
    set(<var> ON)
else()
   set(<var> OFF)
endif()

vcpkg_config_cmake(
    ...
    OPTIONS
        -D<var>=${var}
)
```
So it is time to create a little shiny macro/function for this.

`czmq` has a lot of optional libraries:
```cmake
if("draft" IN_LIST FEATURES)
    set(ENABLE_DRAFTS ON)
else()
    set(ENABLE_DRAFTS OFF)
endif()

if("httpd" IN_LIST FEATURES)
    set(CZMQ_WITH_LIBMICROHTTPD ON)
else()
    set(CZMQ_WITH_LIBMICROHTTPD OFF)
endif()

if("tool" IN_LIST FEATURES)
    set(BUILD_TOOLS ON)
else()
    set(BUILD_TOOLS OFF)
endif()

if("curl" IN_LIST FEATURES)
    set(CZMQ_WITH_LIBCURL ON)
else()
    set(CZMQ_WITH_LIBCURL OFF)
endif()

if("lz4" IN_LIST FEATURES)
    set(CZMQ_WITH_LZ4 ON)
else()
    set(CZMQ_WITH_LZ4 OFF)
endif()

if("uuid" IN_LIST FEATURES AND
   VCPKG_CMAKE_SYSTEM_NAME STREQUAL "Linux")
    set(CZMQ_WITH_UUID ON)
else()
    set(CZMQ_WITH_UUID OFF)
endif()

vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DENABLE_DRAFTS=${ENABLE_DRAFTS}
        -DCZMQ_WITH_LIBMICROHTTPD=${CZMQ_WITH_LIBMICROHTTPD}
        -DCZMQ_WITH_UUID=${CZMQ_WITH_UUID}
        -DCZMQ_WITH_LIBCURL=${CZMQ_WITH_LIBCURL}
        -DCZMQ_WITH_LZ4=${CZMQ_WITH_LZ4}
)
```

This patch provides a macro(vcpkg_check_feature) and a function(vcpkg_check_features). With the help of this new module, it can be simplified as:
```cmake
vcpkg_check_feature(draft ENABLE_DRAFTS)
vcpkg_check_feature(httpd CZMQ_WITH_LIBMICROHTTPD)
vcpkg_check_feature(tool BUILD_TOOLS)
vcpkg_check_feature(lz4 CZMQ_WITH_LZ4)
vcpkg_check_feature(curl CZMQ_WITH_LIBCURL)

if("uuid" IN_LIST FEATURES AND
   VCPKG_CMAKE_SYSTEM_NAME STREQUAL "Linux")
    set(CZMQ_WITH_UUID ON)
else()
    set(CZMQ_WITH_UUID OFF)
endif()

vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DENABLE_DRAFTS=${ENABLE_DRAFTS}
        -DCZMQ_WITH_LIBCURL=${CZMQ_WITH_LIBCURL}
        -DCZMQ_WITH_LIBMICROHTTPD=${CZMQ_WITH_LIBMICROHTTPD}
        -DCZMQ_WITH_LZ4=${CZMQ_WITH_LZ4}
        -DCZMQ_WITH_UUID=${CZMQ_WITH_UUID}
)
```
or
```cmake
vcpkg_check_features(
    draft ENABLE_DRAFTS
    httpd CZMQ_WITH_LIBMICROHTTPD
    tool BUILD_TOOLS
    lz4 CZMQ_WITH_LZ4
    curl CZMQ_WITH_LIBCURL
)

if("uuid" IN_LIST FEATURES AND
   VCPKG_CMAKE_SYSTEM_NAME STREQUAL "Linux")
    set(CZMQ_WITH_UUID ON)
else()
    set(CZMQ_WITH_UUID OFF)
endif()

vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DCZMQ_WITH_UUID=${CZMQ_WITH_UUID}
        ${FEATURE_OPTIONS}
)
```

### CODE REVIEW

**Comment by myd7349** [comment]
I still have the following concerns:
1. Is defining two new functions in one .cmake a good pattern?
2. Is it easy to distinguish between `vcpkg_check_feature` and `vcpkg_check_features`?
3. Is `vcpkg_check_feature(s)` a good name?
4. Does the syntax of `vcpkg_check_features` look weired?

**Comment by Rastaban** [comment]
I'm not sure having two separate macros is needed.  I lean toward just `vcpkg_check_features` (unless others decide it looks weird) because it can still be called with one set of arguments.

**Comment by myd7349** [comment]
@Rastaban Thanks! That sounds resonable to me.

**Comment by myd7349** [comment]
TODO: Take this form into consideration:
```cmake
if(<option>MATCHES "ON")
endif()
```
https://github.com/microsoft/mimalloc/blob/91222691cb21a6fb7a9760a264ff1183a1237fd4/CMakeLists.txt#L45

**Comment by cbezault** [comment]
Agreed with @Rastaban, drop the singular version.

**Comment by myd7349** [comment]
@cbezault Thanks for your reply!

File: scripts/cmake/vcpkg_check_features.cmake [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+## ## Usage
+## ```cmake
+## vcpkg_check_features(
+##     <feature1> <output_variable1>
+##     [<feature2> <output_variable2>]
+##     ...
+## )
+## ```
+##
+## `vcpkg_check_feature` accepts a list of (feature, output_variable) pairs.
```
Comment (cbezault) [inline scripts/cmake/vcpkg_check_features.cmake]: `vcpkg_check_features`

**Comment by cbezault** [comment]
Could you modify one port to use this new feature and add it to this PR?

**Comment by myd7349** [comment]
> 
> 
> Could you modify one port to use this new feature and add it to this PR?

Thanks! Done.

**Comment by myd7349** [comment]
Oops! It seems that I made a mistake here:
```cmake
if("non-posix" IN_LIST FEATURES)
vcpkg_check_features(non-posix ENABLE_POSIX_API)
    set(ENABLE_POSIX_API OFF) # Not `ON` here
else()
    set(ENABLE_POSIX_API ON)
endif()
```
can not be replaced with:
```cmake
vcpkg_check_features(non-posix ENABLE_POSIX_API)
```
I will fix it.

**Comment by myd7349** [comment]
https://github.com/microsoft/vcpkg/pull/7091

**Pull request merged**